module/zfs/zil.c

```diff
@@ -1204,12 +1202,44 @@ zil_lwb_flush_vdevs_done(zio_t *zio)
 
 	mutex_exit(&zilog->zl_lock);
 
-	/*
-	 * Now that we've written this log block, we have a stable pointer
-	 * to the next block in the chain, so it's OK to let the txg in
-	 * which we allocated the next block sync.
-	 */
-	dmu_tx_commit(tx);
+	mutex_enter(&zilog->zl_lwb_io_lock);
+	ASSERT(zilog->zl_lwb_inflight[txg & TXG_MASK] > 0);
+	zilog->zl_lwb_inflight[txg & TXG_MASK]--;
+	if (zilog->zl_lwb_inflight[txg & TXG_MASK] == 0)
+		cv_broadcast(&zilog->zl_lwb_io_cv);
+	mutex_exit(&zilog->zl_lwb_io_lock);
+}
+
+/*
+ * Wait for the completion of all issued write/flush of that txg provided.
+ * It guarantees zil_lwb_flush_vdevs_done() is called and returned.
+ */
+static void
+zil_lwb_flush_wait_all(zilog_t *zilog, uint64_t txg)
+{
+	ASSERT(txg == spa_syncing_txg(zilog->zl_spa));
+
+	mutex_enter(&zilog->zl_lwb_io_lock);
+	while (zilog->zl_lwb_inflight[txg & TXG_MASK] > 0)
+		cv_wait(&zilog->zl_lwb_io_cv, &zilog->zl_lwb_io_lock);
+	mutex_exit(&zilog->zl_lwb_io_lock);
+
+#ifdef ZFS_DEBUG
+	mutex_enter(&zilog->zl_lock);
+	lwb_t *lwb = list_head(&zilog->zl_lwb_list);
+	while (lwb != NULL && lwb->lwb_max_txg <= txg) {
+		if (lwb->lwb_issued_txg <= txg) {
+			ASSERT(lwb->lwb_state != LWB_STATE_ISSUED);
+			ASSERT(lwb->lwb_state != LWB_STATE_WRITE_DONE);
+			IMPLY(lwb->lwb_issued_txg > 0,
+			    lwb->lwb_state == LWB_STATE_FLUSH_DONE);
```

Comment:
> 

Hi amotin, 
I don't have definite answer right now. It has been months since I worked on this PR.

I think here might be the reason.
zil_lwb_flush_wait_all() should only be called for the txg in syncing context. At that point, there will be no new lwbs issued under this txg. In the other words, lwb_issued_txg will not be update to this txg value.

Reply:
No, not this TXG. But if lwb_issued_txg is set to any other in time value between previous `if ()` and this `IMPLY()` (unless compiler cache the value), the last will cause panic.

Reply:
Good catch.
I just pushed 2nd commit.
I added redundant condition in side IMPLY() to make sure the value is still in the range of current txg.

Reply:
I am not sure it makes much sense to me.  Without explicit barriers it relies on some compiler behavior there is nothing guarantied.  Previous version with additional locking had more sense to me, with only one downside that the more debug code becomes different from non-debug, especially in locking, the more chances it it will catch some different issues, not happening in normal builds. ;)

Reply:
I am not sure it makes much sense to me.  Without explicit barriers it relies on some compiler behavior there is nothing guarantied.  Previous version with additional locking had more sense to me, with only one downside that the more debug code becomes different from non-debug, especially in locking, the more chances it it will catch some different issues, not happening in normal builds. ;)

Reply:
I tried to use zl_issuer_lock to protect lwb_issued_txg in zil_lwb_flush_wait_all(). However, there are some existing code paths calling txg_wait_synced() while holding zl_issuer_lock, which let system lock up in zil_lwb_flush_wait_all().
Right now, I wrap lwb_issued_txg inside zl_lwb_io_lock. Please see my UPDATED 2nd commit. Thanks for your advice.